wallet/src/wallet/tx_builder.rs

```diff
@@ -276,29 +276,45 @@ impl<'a, Cs> TxBuilder<'a, Cs> {
     ///
     /// These have priority over the "unspendable" utxos, meaning that if a utxo is present both in
     /// the "utxos" and the "unspendable" list, it will be spent.
+    ///
+    /// If a UTxO is inserted multiple times, only the final insertion will take effect.
     pub fn add_utxos(&mut self, outpoints: &[OutPoint]) -> Result<&mut Self, AddUtxoError> {
-        let utxo_batch = outpoints
+        // Canonicalize once, instead of once for every call to `get_utxo`.
+        let unspent: HashMap<OutPoint, LocalOutput> = self
+            .wallet
+            .list_unspent()
+            .map(|output| (output.outpoint, output))
+            .collect();
+
+        // Ensure that only unique outpoints are added, but keep insertion order.
+        let mut visited = <HashSet<OutPoint>>::new();
+
+        let utxos: Vec<WeightedUtxo> = outpoints
             .iter()
-            .map(|outpoint| {
-                self.wallet
-                    .get_utxo(*outpoint)
-                    .ok_or(AddUtxoError::UnknownUtxo(*outpoint))
-                    .map(|output| {
-                        (
-                            *outpoint,
-                            WeightedUtxo {
-                                satisfaction_weight: self
-                                    .wallet
-                                    .public_descriptor(output.keychain)
-                                    .max_weight_to_satisfy()
-                                    .unwrap(),
-                                utxo: Utxo::Local(output),
-                            },
-                        )
-                    })
+            .map(|&op| -> Result<_, AddUtxoError> {
+                visited.insert(op);
```

Comment:
I still think we should avoid creating more of the same `WeightedUtxo` if the outpoint is already visited.

```suggestion
            .filter(|&&op| visited.insert(op))
            .map(|&op| -> Result<_, AddUtxoError> {
```

For example how should tx builder behave in this situation?

```rs
    let (mut wallet, _) = get_funded_wallet_wpkh();
    let utxo = wallet.list_unspent().next().unwrap();
    let op = utxo.outpoint;

    let mut tx_builder = wallet.build_tx();
    tx_builder.add_utxos(&[op, op, op]).unwrap();

    assert_eq!(tx_builder.params.utxos.len(), 1);
```